sunpy/data/data_manager/cache.py

```diff
@@ -0,0 +1,157 @@
+    storage: Implementation of `~sunpy.data.data_manager.storage.StorageProviderBase`
+        Storage to store metadata about the files.
+    cache_dir: `str` or `pathlib.Path`
+        Directory where the downloaded files will be stored.
+    expiry: `astropy.units.quantity.Quantity` or `None`, optional
+        The interval after which the cache is invalidated. If the expiry is `None`,
+        then the expiry is not checked (or the cache never expires). Defaults to 10 seconds.
+    """
+
+    def __init__(self, downloader, storage, cache_dir, expiry=10*u.s):
+        self._downloader = downloader
+        self._storage = storage
+        self._cache_dir = Path(cache_dir)
+        self._expiry = expiry if expiry is None else TimeDelta(expiry)
+
+    def download(self, urls, redownload=False):
+        """
+        Downloads the files from the urls.
+
+        Parameters
+        ----------
+        urls: `list` or `str`
+            A list of urls or a single url.
+        redownload: `bool`
+            Whether to skip cache and redownload.
+
+        Returns
+        -------
+        `pathlib.PosixPath`
+            Path to the downloaded file.
+        """
+        if isinstance(urls, str):
+            urls = [urls]
+        # Program flow
+        # 1. If redownload: Download, update cache and return file path
+        # 2. If not redownload: Check cache,
+        #    i. If present in cache:
+        #        - If cache expired, remove entry from cache, download and add to cache
+        #        - If cache not expired, return path
+        #    ii. If not download, store in cache and return path
+        details = None
+        for url in urls:
+            details = self._get_by_url(url)
+            if details:
+                break
+        if details:
+            if redownload or self._has_expired(details):
+                # if file is in cache and it has to be redownloaded or the cache has expired
+                # then remove the file and delete the details from the storage
+                os.remove(details['file_path'])
+                self._storage.delete_by_key('url', details['url'])
+            else:
+                return Path(details['file_path'])
+
+        file_path, file_hash, url = self._download_and_hash(urls)
+
+        self._storage.store({
+            'file_hash': file_hash,
+            'file_path': str(file_path),
+            'url': url,
+            'time': datetime.now().isoformat(),
+        })
+        return file_path
+
+    def _has_expired(self, details):
+        """
+        Whether the url corresponding to details in cache has expired or not.
+
+        Parameters
+        ----------
+        details: `dict`
+            Details detched from cache.
+
+        Returns
+        -------
+        `bool`
+            Whether the url has expired or not.
+        """
+        return self._expiry and datetime.now() - datetime.fromisoformat(details['time']) > self._expiry
```

Comment:
Just in case we do want to move this out of core at some point in the future, having astropy as a dependency for just this would be a little over the top?

Reply:
Ah ok. Fair enough